[GSB] Eliminate dependence on specific modules

DougGregor opened the pull request:
Eliminate the `GenericSignatureBuilder` 's dependencies on specific modules:

* The GSB gets its own module-agnostic conformance lookup function (which is the only remaining dependency)
* The module is no longer used to unique canonical generic signature builders (so we only build one canonical GSB per generic signature)
* All of the `ModuleDecl` parameters go away from `GenericSignature`

DougGregor committed:
Always use LookUpConformanceInModule for the GenericSignatureBuilder.

DougGregor committed:
[GSB] Eliminate the stored LookupConformanceFn to the GSB.

Implement a module-agnostic conformance lookup operation within the GSB
itself, so it does not need to be supplied by the code constructing the
generic signature builder. This makes the generic signature builder
(closer to) being module-agnostic.

DougGregor committed:
[AST] Stop uniquing canonical GSBs based on the module.

Now that the GenericSignatureBuilder is no longer sensitive to the input
module, stop uniquing the canonical GSBs based on that module. The main
win here is when deserializing a generic environment: we would end up 
creating a canonical GSB in the module we deserialized and another
canonical GSB in the module in which it is used.

DougGregor committed:
[AST] Eliminate ModuleDecl parameters from GenericSignature.

DougGregor committed:
[AST] Remove ModuleDecl from TypeBase::getCanonicalType().

DougGregor commented:
@swift-ci please test compiler performance

DougGregor commented:
Note that this is currently broken because we're not completing enough conformances (in the type checker) for IRGen.

slavapestov commented:
@dgregor Can LookUpConformanceInSignature go away too?

slavapestov commented on lib/SIL/SILFunctionType.cpp:
```diff
@@ -856,7 +856,7 @@ static CanSILFunctionType getSILFunctionType(SILModule &M,
     auto origGenericSig
       = function->getGenericSignature();
     auto getCanonicalType = [origGenericSig, &M](Type t) -> CanType {
```
No need to capture M anymore

slavapestov replied:
In fact this closure is so trivial I don't think we need it

slavapestov commented on lib/SILGen/SILGenPoly.cpp:
```diff
@@ -2822,9 +2822,8 @@ CanSILFunctionType SILGenFunction::buildThunkType(
                       : ParameterConvention::Direct_Unowned});
 
-  auto &mod = *F.getModule().getSwiftModule();
   auto getCanonicalType = [&](Type t) -> CanType {
```
Also might want to remove this closure too

slavapestov commented:
This is really cool!

How does it break IRGen? The change looks NFC. Is it because GenericSignatureBuilder::lookupConformance() no longer goes through TypeChecker::conformsToProtocol and the conformance is never "forced"?

slavapestov commented:
Also what about the places that use a ModuleDecl to recover conformances when building substitutions from a BoundGenericType? Could those in theory use the canonical GSB's lookupConformance() routine instead?

DougGregor commented:
Your intuition about `TypeChecker::conformsToProtocol` is correct; that's what's currently forcing conformances to be complete, and it's not happening as often now.

DougGregor commented:
The other entry point you're referring to is... `TypeBase::getContextSubstitutionMap()`, maybe? It seems like we really need to have `BoundGenericType` store the actual substitutions (at least in its fully-type-checked form), because those could be module-dependent.

DougGregor commented:
(I'll remove the closures in a follow-up commit)

slavapestov commented:
@DougGregor Would it make sense for getSubstitutionMap() to use the BoundGenericType's signature's canonical GSB for the lookups?

DougGregor committed:
Minor simplifications suggested by Slava

DougGregor commented:
A `BoundGenericType` should already have all of the substitutions it needs.

DougGregor committed:
[Cleanup] Reduce dependencies on swift/AST/LazyResolver.h.

DougGregor committed:
[GSB] Mark conformances used by the GenericSignatureBuilder as “used”.

Prior to the removal of the “lookup conformance function”, whenever
the type checker created a GenericSignatureBuilder, its lookup conformance
function would mark any referenced conformance as “used”. Do so now via
a LazyResolver callback, which fixes a regression in multi-file type
checking scenarios.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
Yeah, I pushed without building locally after resolving a merge conflict.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge
